### PR TITLE
Feat/dApp signer enhancement

### PR DESCRIPTION
**Pull request by franfernandez20**
**Description**:
General improvements to the DAppConnector.  The enhancements cover various aspects of the connector, supporting several sessions, and facilitating the use of the client from the dApps side.

* Add methods to directly use transaction
* Support multiple sessions
* Remove unnecessary code 
* Enhance certain performance aspects.

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: README.md
Comment (zanctor) [inline README.md:181]: I guess, this should be a string.

File: src/lib/dapp/index.ts
Comment (zanctor) [inline src/lib/dapp/index.ts:71]: Is there a reason why the modal instance should be a connector property?

Comment (franfernandez20) [reply src/lib/dapp/index.ts:71]: To support the connect functionality through modals. And also is exposed to be able to manage posible events. 
Like, ```walletConnectModal.subscribeModal```

**Comment by franfernandez20** [comment]
Hi @mgarbs, can I kindly request to have another reviewer assigned to this PR?

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:160]: call needs to be implemented to support `transaction.executeWithSigner(signer)` https://github.com/hashgraph/hedera-sdk-js/blob/65a2323f8d9281d686c2f1c4d337b7a22a3ca97c/src/Executable.js#L500

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:164]: I see how these methods could be useful to dApp developers. But exposing these methods may back walletconnect into a corner where it's Signer implementation exposes extra methods that dApps become dependent. I think these extra methods should all be private, unless Signer also publicly exposes the same methods.

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:65]: `this.nodesAccountIds = [AccountId.fromString('0.0.3')]` is a single point of failure on consensus node 3. Consider creating a hashgraph/sdk client and grabbing node ids from it's _network property. See this Signer implementation from the hashgraph/sdk https://github.com/hashgraph/hedera-sdk-js/blob/main/src/Wallet.js#L272

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:78]: Same comment. I think nodeAccountIds can be fetched more reliably using a `Client` or `Provider` instead of allowing the dApp to explicitly set them like this.

If the dApp wants to explicitly set nodeAccountIds, it should do so on the transaction itself instead of DAppSigner.

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:68]: I think it was more elegant to pass the signClient in the constructor like it was before instead of using a static method to set a static prop that's shared between all instances.

File: src/lib/dapp/index.ts
Comment (pwoosam) [inline src/lib/dapp/index.ts:154]: nit: Returning undefined should be ok if not found. The dApp is likely better off handling an undefined result than try/catching this exception.

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [inline src/lib/dapp/DAppSigner.ts:52]: Should props that are not already exposed by the Signer interface be private?

File: src/lib/dapp/index.ts
Comment (pwoosam) [inline src/lib/dapp/index.ts:193]: Excellent catch with this finally block!

File: src/lib/dapp/index.ts
Comment (pwoosam) [inline src/lib/dapp/index.ts:372]: I'm questioning if we should remove the execute*/sign* methods from the DAppConnector entirely and instead only use the Signer for executing transactions.

It feels like the DAppConnector should limit its responsibility managing walletconnect sessions and extending walletconnect.

**Review by pwoosam — CHANGES_REQUESTED** [review]

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:350]: Yes, the problem with that code is that it removes the session if the wallet is not replying. We have the disconnect or disconnectAll methods.

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:389]: Completely agree with that @justynspooner but that is opinionable and a matter of preferences in the coding.

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:501]: I don't agree. I think the library should be as raw as possible. However, a new PR to discuss it would be great.

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:128]: Yes, absolutely. I did as much as I could. Let's unlock this contribution and we can continue improving it.

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:164]: If you want to use separately the signer functionality from the DAppConnector and session manage it is mandatory to keep those methods public. Furthermore, there are no more issues to having them exposed than having nothing to use.

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:65]: With the sdk client, you don't ensure you use the same nodes as the wallet you are using. So I think the best option is to set manually the available nodes after requesting them to the wallet. Anyway, I think here we have a good point to improve. I suggest creating a new PR to address the issue.

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:78]: Same suggestion :)

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:68]: That is debatable. For now, we have nothing to ease the dApp developers' integration. Let's start with this contribution and then discuss minor details like this.

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:193]: Instead of being sarcastic and nitpicking, please be proactive and offer solutions. Also, it would be great if your comments included code solutions. Like other colleagues do.

Comment (pwoosam) [reply src/lib/dapp/index.ts:193]: I'm sorry for the misunderstanding :( I was genuine with this comment. When errors were thrown previously, the modal stayed open and was unusable. This was an excellent find.

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [reply src/lib/dapp/DAppSigner.ts:65]: > With the sdk client, you don't ensure you use the same nodes as the wallet you are using.
I don't think this should be a requirement for walletconnect. The `nodeAccountIds` on a transaction can differ from the nodes returned by a wallet, as long as they are nodes for the same network.

**Comment by pwoosam** [comment]
@franfernandez20  I don't support merging this PR in its current state. The PR comments I left were marked resolved without any changes.

To unblock this PR, I think it should be split into at least 3 PRs with minimal changes:
- DAppSigner enhancement
- Close walletConnectModal on error
- DAppConnector execution/sign method improvements

A PR should have a single, unique goal. This way contributors can agree on the minimal set of changes and focus discussion on a single topic to unblock the PR.

File: src/lib/dapp/DAppSigner.ts
Comment (teacoat) [reply src/lib/dapp/DAppSigner.ts:65]: @franfernandez20 we absolutely cannot be hardcoding node id's like this, if node 3 goes down all transactions using this node will fail

File: src/lib/dapp/DAppSigner.ts
Comment (pwoosam) [reply src/lib/dapp/DAppSigner.ts:164]: > If you want to use separately the signer functionality from the DAppConnector and session manage it is mandatory to keep those methods public.

I'm not completely understanding this. Can you explain what you mean by `use separately the signer functionality from the DAppConnector and session manage`?

My understanding is that `DAppConnector` should be used to session manage and `DAppSigner` should be used as an implementation for the hashgraph/sdk's Signer interface. Right now `DAppSigner` is leaking its abstraction details to the user.

**Comment by pwoosam** [comment]
I decided to take a shot at the DAppSigner enhancement in another PR: #116

**Comment by hgraphql** [comment]
It looks like there's a bit of unresolved comments that have been marked resolved. Moving forward can we have both the commenter and the contributor mark a conversation as resolved?

**Comment by hgraphql** [comment]
> I decided to take a shot at the DAppSigner enhancement in another PR: #116

Should we view this as an alternate approach or a complementary PR?

Are there pieces that overlap?

We may want to consider a "release" branch so that we can begin to come to consensus on a more granular basis.

**Comment by pwoosam** [comment]
@hgraphql #116 is an alternate approach.

#116 implements more of Signer interface, enables transaction.executeWithSigner(signer) usage, adds unit tests for the signer, updates the example dApp have new sections to test the signer methods end to end, addresses the feedback in this PR, and adds support for some queries that return array results instead of individual objects.

**Comment by pwoosam** [comment]
There is significant overlap between the PRs, but #116 focuses on only what is needed to support implementing the Signer interface and testing it. #76 has a few other unrelated enhancements in it.

File: src/lib/dapp/DAppSigner.ts
Comment (franfernandez20) [reply src/lib/dapp/DAppSigner.ts:65]: To set the list, you can use the setNodes function. It is essential to ensure that the nodes in the client and wallet are aligned. Although we need to restore the Transaction.toBytes() approach to retain more nodes, we now have to use this method with only one node. 

Anyway, what do you propose?

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:154]: Not agree

File: src/lib/dapp/index.ts
Comment (franfernandez20) [reply src/lib/dapp/index.ts:372]: I agree with that but I keep there to be backwards compatible with the previous versions.

**Comment by franfernandez20** [comment]
> @franfernandez20 I don't support merging this PR in its current state. The PR comments I left were marked resolved without any changes.
> 
> To unblock this PR, I think it should be split into at least 3 PRs with minimal changes:
> 
> * DAppSigner enhancement
> * Close walletConnectModal on error
> * DAppConnector execution/sign method improvements
> 
> A PR should have a single, unique goal. This way contributors can agree on the minimal set of changes and focus discussion on a single topic to unblock the PR.

Indeed, it would be preferable to have fewer PRs. However, we are starting from an unfinished version of this dAppSigner, which complicates the process of having one PR per topic, especially since there are more than three topics involved. While it might require more effort, I will attempt to manage it. Nonetheless, I believe it might be more efficient to first achieve a functional version and then focus on making improvements.

**Comment by hgraphql** [comment]
Great convos here!

 Closing this PR in favor #116 , #134 , #135 
 
 If I'm mistaken on this feel free to re-open!

**Pull request closed without merging**
